astroid/constraint.py

```diff
@@ -0,0 +1,135 @@
+# Licensed under the LGPL: https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html
+# For details: https://github.com/PyCQA/astroid/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
+
+"""Classes representing different types of constraints on inference values."""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TypeVar, Union
+
+from astroid import nodes, util
+
+_NameNodes = Union[nodes.AssignAttr, nodes.Attribute, nodes.AssignName, nodes.Name]
+_ConstraintT = TypeVar("_ConstraintT", bound="Constraint")
+
+
+class Constraint(ABC):
+    """Represents a single constraint on a variable."""
+
+    def __init__(self, node: nodes.NodeNG, negate: bool) -> None:
+        self.node = node
+        """The node that this constraint applies to."""
+        self.negate = negate
+        """True if this constraint is negated. E.g., "is not" instead of "is"."""
+
+    @classmethod
+    @abstractmethod
+    def match(
+        cls: _ConstraintT, node: _NameNodes, expr: nodes.NodeNG, negate: bool = False
+    ) -> _ConstraintT | None:
+        """Return a new constraint for node matched from expr, if expr matches
+        the constraint pattern.
+
+        If negate is True, negate the constraint.
+        """
+
+    @abstractmethod
+    def satisfied_by(self, inferred: nodes.NodeNG | type[util.Uninferable]) -> bool:
+        """Return True if this constraint is satisfied by the given inferred value."""
+
+
+class NoneConstraint(Constraint):
+    """Represents an "is None" or "is not None" constraint."""
+
+    CONST_NONE: nodes.Const = nodes.Const(None)
+
+    @classmethod
+    def match(
+        cls: _ConstraintT, node: _NameNodes, expr: nodes.NodeNG, negate: bool = False
+    ) -> _ConstraintT | None:
+        """Return a new constraint for node matched from expr, if expr matches
+        the constraint pattern.
+
+        Negate the constraint based on the value of negate.
+        """
+        if isinstance(expr, nodes.Compare) and len(expr.ops) == 1:
+            left = expr.left
+            op, right = expr.ops[0]
+            if op in {"is", "is not"} and (
+                _matches(left, node)
+                and _matches(right, cls.CONST_NONE)
+                or _matches(left, cls.CONST_NONE)
```

Comment:
`cls.CONST_NONE` is a type error. See: https://github.com/PyCQA/astroid/pull/1189#discussion_r903136395

Reply:
I think this might be fixed with using `Self`?